plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ScanQueryPageSource.java

```diff
@@ -267,67 +265,59 @@ private void flattenFields(Map<String, Type> result, String fieldName, Type type
                         return new SourceColumnDecoder();
                     }
 
-                    return createDecoder(column.getName(), column.getType());
+                    return createDecoder(column.getName(), column.getDecoderContext());
                 })
                 .collect(toImmutableList());
     }
 
-    private Decoder createDecoder(String path, Type type)
+    private Decoder createDecoder(String path, DecoderContext decoderContext)
     {
-        if (type.equals(VARCHAR)) {
-            return new VarcharDecoder(path);
+        switch (decoderContext.getDecoderType()) {
+            case RAW_JSON:
+                return new RawJsonDecoder(path);
+            case VARCHAR:
+                return new VarcharDecoder(path);
+            case VARBINARY:
+                return new VarbinaryDecoder(path);
+            case TIMESTAMP:
+            case TIMESTAMP_WITH_TIME_ZONE:
+                return new TimestampDecoder(path);
+            case BOOLEAN:
+                return new BooleanDecoder(path);
+            case DOUBLE:
+                return new DoubleDecoder(path);
+            case REAL:
+                return new RealDecoder(path);
+            case TINYINT:
+                return new TinyintDecoder(path);
+            case SMALLINT:
+                return new SmallintDecoder(path);
+            case INTEGER:
+                return new IntegerDecoder(path);
+            case BIGINT:
+                return new BigintDecoder(path);
+            case IP_ADDRESS:
+                return new IpAddressDecoder(path, ipAddressType);
+            case ROW:
+                Map<String, DecoderContext> children = decoderContext.getChildren();
+
+                List<Decoder> decoders = children.values().stream()
+                        .map(child -> createDecoder(
+                                appendPath(path, child.getName()),
+                                child))
+                        .collect(toImmutableList());
+
+                List<String> fieldNames = children.values().stream()
+                        .map(DecoderContext::getName)
+                        .collect(toImmutableList());
+
+                return new RowDecoder(path, fieldNames, decoders);
```

Comment:
The impl is a `LinkedHashmap` so order is sustained for now but I do agree this is a bit buggy if we don't enforce the Map class. I still think the main idea @martint was pushing before was to totally separate the type from the decoder implementation outside of the `getDecoderContext()` method. 

These maps are created in `DecoderContext`, what if we push the creation and management of these types into `DecoderContext` and enforce an API that internally uses `LinkedHashMap` and exposes `addChild(DecoderContext context)` then `getChildrenDecoders()` and `getFieldNames()`.

Reply:
I mean another really simple change would be:
```
                List<DecoderContext> = decoderContext.getChildren().values();

                List<Decoder> decoders = children.stream()
                        .map(child -> createDecoder(
                                appendPath(path, child.getName()),
                                child))
                        .collect(toImmutableList());

                List<String> fieldNames = children.stream()
                        .map(DecoderContext::getName)
                        .collect(toImmutableList());
```

Reply:
I agree. While the implementation is a `LinkedHashmap`, the interface is `Map`, and as such, it doesn't guarantee ordering. We will need some way to encode the order of the fields.

Reply:
I think we should still keep using `RowType`'s field-ordering for this, since that's how we propagate nested block ordering in connectors like hive and iceberg as well. 

The decoder context will be simpler that way, and doesn't need to know any ordering.